### PR TITLE
Support onion message pathfinding 

### PR DESCRIPTION
**Pull request by valentinewallace**
We *could* reuse router::get_route, but it's better to start from scratch
because get_route includes a lot of payment-specific computations that impact
performance.

Partially addresses #1607

### CODE REVIEW

File: lightning/src/onion_message/router.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	FinalHopConversion,
+}
+
+//  Heavily adapted from https://github.com/samueltardieu/pathfinding/blob/master/src/directed/dijkstra.rs
+//  TODO: how2credit the repo (is that necessary?)?
+/// Run Dijkstra's from `start` until `found_target` indicates that we've found the destination.
+/// `successor_callback` must invoke the callback that it is provided on each of a given node's
+/// next-hop peers. `convert_final_hop` may be used to convert an intermediate processing hop type
+/// (`N`) to a final path hop type (`H`).
+fn dijkstra<N, H, FN, FS, FC>(start: N, successor_callback: &mut FN, found_target: FS,
```
Comment (tnull) [inline lightning/src/onion_message/router.rs:89]: This interface seems to be kept very generic to accommodate different usage scenarios by the original author. I think when we're adapting this we can fill in a lot of these types, which should improve readability?

File: lightning/src/onion_message/router.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	}
+}
+
+// Errors if `convert_path_hop` fails.
+fn reverse_path<N, H, FC>(parents: HashMap<N, (N, u64)>, start: N, convert_path_hop: &mut FC) -> Result<Vec<H>, ()>
+	where N: Eq + Hash + Copy + Ord,
+				FC: FnMut(&N) -> Result<H, ()>,
+{
+	let mut path = vec![convert_path_hop(&start)?];
+	let mut curr = start;
```
Comment (tnull) [inline lightning/src/onion_message/router.rs:142]: IIUC, this `start` is actually the target hop. So renaming the variable accordingly could spare future confusion.

**Comment by valentinewallace** [comment]
Ok it's clearly less readable than I thought with a generic Dijkstra's. Staying generic would allow us to pretty much reuse the test vectors in the original repo, but it doesn't seem worth. Gonna convert to draft for now and rewrite the generic portions

**Comment by tnull** [comment]
> Ok it's clearly less readable than I thought with a generic Dijkstra's. Staying generic would allow us to pretty much reuse the test vectors in the original repo, but it doesn't seem worth. Gonna convert to draft for now and rewrite the generic portions

Ah, didn't realize keeping it generic had that intention. Well, it's not *unreadable* and reusing test vectors may be worth it. 
If we however also see a way of refactoring the payment pathfinding to reuse the same generic implementation, we'd should really consider it, IMO.

**Comment by valentinewallace** [comment]
> I believe the upstream dijkstras you pulled from is actually buggy, anyway, so the generic-ness isn't worth much :p

Lol I might've introduced bugs in my quest to eliminate all vecs from the original, to be fair

**Comment by valentinewallace** [comment]
> > Ok it's clearly less readable than I thought with a generic Dijkstra's. Staying generic would allow us to pretty much reuse the test vectors in the original repo, but it doesn't seem worth. Gonna convert to draft for now and rewrite the generic portions
> 
> Ah, didn't realize keeping it generic had that intention. Well, it's not _unreadable_ and reusing test vectors may be worth it. If we however also see a way of refactoring the payment pathfinding to reuse the same generic implementation, we'd should definitely consider it, IMO.

I did wonder if there were other cases where we'd want a generic implementation. I have no idea whether that'd be realistic with our payment pathfinding though lol, we'd need a way to "decrement available liquidity" for hops in paths that were already found, at least, but maybe that's easy? 🤔

File: lightning/src/onion_message/router.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+use alloc::collections::BinaryHeap;
+use core::hash::Hash;
+use core::ops::Deref;
+use prelude::*;
+
+/// Find a path for sending an onion message.
+pub fn find_path<L: Deref, GL: Deref>(
+	our_node_pubkey: &PublicKey, receiver_pubkey: &PublicKey, network_graph: &NetworkGraph<GL>, first_hops: Option<&[&PublicKey]>, logger: L
+) -> Result<Vec<PublicKey>, LightningError> where L::Target: Logger, GL::Target: Logger
```
Comment (ariard) [inline lightning/src/onion_message/router.rs:15]: So we wouldn't reuse all the information we're learning and storing in our `ProbabilisticScorer`. I can see how we're limited with the current penalty being based on the link-level and here we might be interested by node-level reliability in the path construction. That said, we can also assume that a reliable channel == a reliable onion message communication channel and go with it. I don't know if the spec says anything here or what are the thinking of other implementations ?

**Comment by valentinewallace** [comment]
Thanks for all the feedback! This needs a lot of work and I'm switching it up to work on custom OMs first -- closing for now.

File: lightning/src/onion_message/router.rs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+enum Error {
+	/// No path exists to the destination.
+	PathNotFound,
+	/// Converting the processing hop type to the final hop type failed, see `dijkstra`'s
+	/// `convert_final_hop` parameter.
+	FinalHopConversion,
+}
+
+//  Heavily adapted from https://github.com/samueltardieu/pathfinding/blob/master/src/directed/dijkstra.rs
+//  TODO: how2credit the repo (is that necessary?)?
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/onion_message/router.rs:84]: #1724 ended up not adapting this crate's implementation anymore (shoutout to Wikpedia)

**Pull request closed without merging**